File: sdk/src/ended.ts
/**
 * Connection ends: the vault's refusals and status answers turned into
 * ConnectionEndedError, and the page told. Every end passes through
 * ended(), so a mounted Connect your AI button hears about it without the
 * app writing a line: the error is dispatched on the document as an
 * "outlet-connection-ended" event and the button bound to that grant shows
 * the screen for the reason. A runtime without a document (a server, a
 * worker) has no button to tell and dispatches nothing.
 */
import { getProvider } from "./providers.js";
import {
  ConnectionEndedError, type EndReason, type GrantInfo, OutletError, type Provider,
} from "./types.js";

export const ENDED_EVENT = "outlet-connection-ended";

function announce(e: ConnectionEndedError): void {
  if (typeof document === "undefined" || typeof CustomEvent === "undefined") return;
  document.dispatchEvent(new CustomEvent(ENDED_EVENT, { detail: e }));
}

/** The typed error, announced to the page. */
export function ended(
  reason: EndReason,
  grantId: string,
  extra: { info?: GrantInfo; provider?: Provider; status?: number } = {},
): ConnectionEndedError {
  const message = reason === "refused" && extra.provider
    ? `${getProvider(extra.provider)?.displayName ?? extra.provider} refused this Direct API key. Ask the user for a new one.`
    : undefined;
  const e = new ConnectionEndedError(reason, grantId, { ...extra, ...(message ? { message } : {}) });
  announce(e);
  return e;
}

/** The connection end behind a vault refusal, or null when the code is
 *  not one: 409 grant_capped, 409 grant_revoked, and 401 unauthorized for
 *  a refresh token the vault no longer knows (rotated away or gone). */
export function endedFromVault(e: OutletError, grantId: string): ConnectionEndedError | null {
  if (e.code === "grant_capped") return ended("capped", grantId, { status: e.status });
  if (e.code === "grant_revoked") return ended("revoked", grantId, { status: e.status });
  if (e.code === "unauthorized") return ended("expired", grantId, { status: e.status });
  return null;
}

/** The connection end a status answer reports, or null while it is open. */
export function endedFromInfo(info: GrantInfo): ConnectionEndedError | null {
  if (info.status !== "capped" && info.status !== "revoked") return null;
  return ended(info.status, info.grantId, { info, provider: info.providers[0], status: 409 });
}

/** True for a ConnectionEndedError from any copy of the SDK on the page. */
export function isConnectionEnded(e: unknown): e is ConnectionEndedError {
  if (e instanceof ConnectionEndedError) return true;
  const x = e as { code?: unknown; reason?: unknown; grantId?: unknown } | null;
  return !!x && typeof x === "object" && x.code === "connection_ended"
    && typeof x.reason === "string" && typeof x.grantId === "string";
}

File: sdk-python/src/useoutlet/ended.py
"""Connection ends: the vault's refusals and status answers turned into
ConnectionEndedError. Every end passes through ended(). The npm package also
tells the page's Connect your AI button; a server has no button to tell."""

from __future__ import annotations

from .providers import get_provider
from .types import ConnectionEndedError, EndReason, GrantInfo, OutletError


def ended(
    reason: EndReason,
    grant_id: str,
    *,
    info: GrantInfo | None = None,
    provider: str | None = None,
    status: int | None = None,
) -> ConnectionEndedError:
    """The typed error for a connection that ended."""
    message = None
    if reason == "refused" and provider:
        entry = get_provider(provider)
        name = entry.display_name if entry else provider
        message = f"{name} refused this Direct API key. Ask the user for a new one."
    return ConnectionEndedError(
        reason, grant_id, info=info, provider=provider, status=status, message=message
    )


def ended_from_vault(e: OutletError, grant_id: str) -> ConnectionEndedError | None:
    """The connection end behind a vault refusal, or None when the code is not
    one: 409 grant_capped, 409 grant_revoked, and 401 unauthorized for a refresh
    token the vault no longer knows (rotated away or gone)."""
    if e.code == "grant_capped":
        return ended("capped", grant_id, status=e.status)
    if e.code == "grant_revoked":
        return ended("revoked", grant_id, status=e.status)
    if e.code == "unauthorized":
        return ended("expired", grant_id, status=e.status)
    return None


def ended_from_info(info: GrantInfo) -> ConnectionEndedError | None:
    """The connection end a status answer reports, or None while it is open."""
    if info.status not in ("capped", "revoked"):
        return None
    provider = info.providers[0] if info.providers else None
    return ended(info.status, info.grant_id, info=info, provider=provider, status=409)


def is_connection_ended(e: object) -> bool:
    """True for a ConnectionEndedError, from this package or another copy of it."""
    if isinstance(e, ConnectionEndedError):
        return True
    return (
        getattr(e, "code", None) == "connection_ended"
        and isinstance(getattr(e, "reason", None), str)
        and isinstance(getattr(e, "grant_id", None), str)
    )
